src/mqt/predictor/ml/predictor.py

```diff
@@ -210,58 +220,60 @@ def generate_training_sample(
             raise RuntimeError("File is not a qasm file: " + str(file))
 
         logger.debug("Checking " + str(file))
-        scores = [-1.0 for _ in range(len(self.devices))]
+        scores = {dev.name: -1.0 for dev in self.devices}
         all_relevant_files = path_compiled_circuits.glob(str(file).split(".")[0] + "*")
 
         for filename in all_relevant_files:
             filename_str = str(filename)
-            if (str(file).split(".")[0] + "_" + figure_of_merit + "_") not in filename_str and filename_str.endswith(
+            if (str(file).split(".")[0] + "_" + self.figure_of_merit) not in filename_str and filename_str.endswith(
                 ".qasm"
             ):
                 continue
-            comp_path_index = int(filename_str.split("_")[-1].split(".")[0])
-            device = self.devices[comp_path_index]
+            dev_name = filename_str.split("-")[-1].split(".")[0]
+            if dev_name not in [dev.name for dev in self.devices]:
+                continue
+            device = get_device_by_name(dev_name)
             qc = QuantumCircuit.from_qasm_file(filename_str)
-            if figure_of_merit == "critical_depth":
+            if self.figure_of_merit == "critical_depth":
                 score = reward.crit_depth(qc)
-            elif figure_of_merit == "expected_fidelity":
+            elif self.figure_of_merit == "expected_fidelity":
                 score = reward.expected_fidelity(qc, device)
-            elif figure_of_merit == "estimated_success_probability":
+            elif self.figure_of_merit == "estimated_success_probability":
                 score = reward.estimated_success_probability(qc, device)
             else:
-                assert_never(figure_of_merit)
-            scores[comp_path_index] = score
+                assert_never(self.figure_of_merit)
+            scores[dev_name] = score
 
         num_not_empty_entries = 0
-        for i in range(len(self.devices)):
-            if scores[i] != -1.0:
+        for dev in self.devices:
+            if scores[dev.name] != -1.0:
                 num_not_empty_entries += 1
 
         if num_not_empty_entries == 0:
             logger.warning("no compiled circuits found for:" + str(file))
 
         feature_vec = ml.helper.create_feature_dict(path_uncompiled_circuit / file)
-        training_sample = (list(feature_vec.values()), np.argmax(scores))
+        target_label = np.argmax(list(scores.values()))
+
+        training_sample = (list(feature_vec.values()), target_label)
         circuit_name = str(file).split(".")[0]
-        return training_sample, circuit_name, scores
+        return training_sample, circuit_name, list(scores.values())
```

Comment:
you are converting `scores` to a list twice here. might be worth to just do this once and store the result in a variable.

Reply:
True, I improved that.